src/order/filter/n_ary.lean

```diff
@@ -155,104 +109,74 @@ lemma map₂_inf_subset_right : map₂ m f (g₁ ⊓ g₂) ≤ map₂ m f g₁ 
 le_inf (map₂_mono_left inf_le_left) (map₂_mono_left inf_le_right)
 
 @[simp] lemma map₂_pure_left : map₂ m (pure a) g = g.map (λ b, m a b) :=
-filter.ext $ λ u, ⟨λ ⟨s, t, hs, ht, hu⟩,
-  mem_of_superset (image_mem_map ht) ((image_subset_image2_right $ mem_pure.1 hs).trans hu),
-    λ h, ⟨{a}, _, singleton_mem_pure, h, by rw [image2_singleton_left, image_subset_iff]⟩⟩
+by simp only [← map_prod_eq_map₂, pure_prod, map_map]
 
 @[simp] lemma map₂_pure_right : map₂ m f (pure b) = f.map (λ a, m a b) :=
-filter.ext $ λ u, ⟨λ ⟨s, t, hs, ht, hu⟩,
-  mem_of_superset (image_mem_map hs) ((image_subset_image2_left $ mem_pure.1 ht).trans hu),
-    λ h, ⟨_, {b}, h, singleton_mem_pure, by rw [image2_singleton_right, image_subset_iff]⟩⟩
+by simp only [← map_prod_eq_map₂, prod_pure, map_map]
 
 lemma map₂_pure : map₂ m (pure a) (pure b) = pure (m a b) := by rw [map₂_pure_right, map_pure]
 
 lemma map₂_swap (m : α → β → γ) (f : filter α) (g : filter β) :
   map₂ m f g = map₂ (λ a b, m b a) g f :=
-by { ext u, split; rintro ⟨s, t, hs, ht, hu⟩; refine ⟨t, s, ht, hs, by rwa image2_swap⟩ }
+by { ext u, split; rintro ⟨s, hs, t, ht, hu⟩; refine ⟨t, ht, s, hs, by rwa image2_swap⟩ }
 
 @[simp] lemma map₂_left (h : g.ne_bot) : map₂ (λ x y, x) f g = f :=
-begin
-  ext u,
-  refine ⟨_, λ hu, ⟨_, _, hu, univ_mem, (image2_left $ h.nonempty_of_mem univ_mem).subset⟩⟩,
-  rintro ⟨s, t, hs, ht, hu⟩,
-  rw image2_left (h.nonempty_of_mem ht) at hu,
-  exact mem_of_superset hs hu,
-end
+by rw [← map_prod_eq_map₂, map_fst_prod]
 
 @[simp] lemma map₂_right (h : f.ne_bot) : map₂ (λ x y, y) f g = g := by rw [map₂_swap, map₂_left h]
 
+lemma map_map₂ (m : α → β → γ) (n : γ → δ) : (map₂ m f g).map n = map₂ (λ a b, n (m a b)) f g :=
+by simp only [← map_prod_eq_map₂, map_map]
+
+lemma map₂_map_left (m : γ → β → δ) (n : α → γ) :
+  map₂ m (f.map n) g = map₂ (λ a b, m (n a) b) f g :=
+by { rw [← map_prod_eq_map₂, ← map_prod_eq_map₂, prod_map_left, map_map], refl }
+
+lemma map₂_map_right (m : α → γ → δ) (n : β → γ) :
+  map₂ m f (g.map n) = map₂ (λ a b, m a (n b)) f g :=
+by rw [map₂_swap, map₂_map_left, map₂_swap]
+
+/-! ### `filter.map₃` -/
```

Comment:
I'm not sure those headers are terribly relevant given that `filter.map₃` is an implementation detail to prove more `filter.map₂` lemmas.